src/main/kotlin/com/beust/kobalt/maven/MavenId.kt

```diff
@@ -10,43 +10,52 @@ package com.beust.kobalt.maven
  * This class accepts a versionless id, which needs to end with a :, e.g. "com.beust:jcommander:" (which
  * usually means "latest version") but it doesn't handle version ranges yet.
  */
-public class MavenId(val id: String) {
-    lateinit var groupId: String
-    lateinit var artifactId: String
-    var packaging: String? = null
-    var version: String? = null
+data class MavenId(val groupId: String, val artifactId: String, val version: String?, val packaging: String? = null, val classifier: String? = null) {
 
     companion object {
+        @Deprecated("", ReplaceWith("MavenId(groupId, artifactId, version, packaging)", "com.beust.kobalt.maven.MavenId"))
         fun create(groupId: String, artifactId: String, packaging: String?, version: String?) =
-               MavenId(toId(groupId, artifactId, packaging, version))
+                MavenId(groupId, artifactId, version, packaging)
 
-        fun toId(groupId: String, artifactId: String, packaging: String? = null, version: String?) =
-                "$groupId:$artifactId" +
-                    (if (packaging != null) ":$packaging" else "") +
-                    ":$version"
+        fun toId(groupId: String, artifactId: String, packaging: String? = null, version: String?, classifier: String? = null) =
+                listOf(groupId, artifactId, packaging, version, classifier).filterNotNull().joinToString(":")
     }
 
-    init {
-        val c = id.split(":")
-        if (c.size != 3 && c.size != 4) {
-            throw IllegalArgumentException("Illegal id: $id")
-        }
-        groupId = c[0]
-        artifactId = c[1]
-        if (! c[2].isEmpty()) {
-            if (isVersion(c[2])) {
-                version = c[2]
-            } else {
-                packaging = c[2]
-                version = c[3]
-            }
-        }
-    }
+    val hasVersion: Boolean
+        get() = version != null
+
+    val id: String
+        get() = MavenId.toId(groupId, artifactId, packaging, version, classifier)
 
-    private fun isVersion(s: String) : Boolean = Character.isDigit(s[0])
+}
 
-    val hasVersion = version != null
+fun MavenId(id: String): MavenId {
+    val c = id.split(":").toLinkedList()
 
-    val toId = MavenId.toId(groupId, artifactId, packaging, version)
+    if (c.size !in 2 .. 5) {
```

Comment:
We really need tests for all the combinations that a `MavenId` can have now, this is becoming hard to follow.